// fluid-configurations/Gruntfile.js
module.exports = function (grunt) {

  require("load-grunt-tasks")(grunt);
  require("time-grunt")(grunt);

  // Project configuration.
  grunt.initConfig({
    pkg: grunt.file.readJSON("package.json"),
    jshint: {
      files: ["Gruntfile.js", "lib/*.js", "test/*.js", "test/**/*.js"],
      options: {
        jshintrc: "./.jshintrc"
      }
    },
    watch: {
      all: {
        files: ["Gruntfile.js", "lib/*.js", "test/*.js", "test/**/*.js"],
        tasks: ["jshint", "exec:touch"]
      },
      man: {
        file: ["docs/cli.md"],
        tasks: ["markedman"]
      }
    },
    release: {
      options: {
        npm: false
      }
    },
    exec: {
      touch: {
        command: "touch package.json"
      }
    },
    open: {
      coverage: {
        path: "./coverage.html"
      }
    },
    markedman: {
      options: {
        version: "<%= pkg.version %>",
        section: "1"
      },
      readme: {
        files: [
          { src: "docs/cli.md", dest: "man/fluid-configurations.1" }
        ]
      }
    }
  });

  // Default task.
  grunt.registerTask("default", "watch");

};
